### PR TITLE
Go asset builder

### PR DESCRIPTION
**Pull request by jvatic**
Replaces `util/rubyassetbuilder` with `util/goassetbuilder` for building installer and dashboard assets, resulting in 2x+ faster build times through parallelization.

This still uses ruby as there are a few erb files, and calls out to nodejs for transforming ES6 modules and SCSS.

### CODE REVIEW

File: installer/app/compiler.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
+package main
+
+import (
+	"encoding/json"
+	"log"
+	"os"
+	"path/filepath"
+
+	matrix "github.com/jvatic/asset-matrix-go"
```
Comment (titanous) [inline installer/app/compiler.go]: This needs to be vendored with godep.

**Comment by titanous** [comment]
I think we can just call the new one `util/assetbuilder` since everything is Go.

**Comment by jvatic** [comment]
@titanous comments addressed

File: util/assetbuilder/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+# install Go
+RUN wget https://storage.googleapis.com/golang/go1.4.2.linux-amd64.tar.gz
+RUN tar -C /usr/local -zxvf go1.4.2.linux-amd64.tar.gz
+ENV PATH /usr/local/go/bin:$PATH
+ENV GOPATH /build/gopath
+
+RUN curl -sL https://deb.nodesource.com/setup | bash -
+
+RUN apt-get update &&\
+    apt-get install -y nodejs build-essential &&\
```
Comment (titanous) [inline util/assetbuilder/Dockerfile]: apt-get and curl install node?

File: util/assetbuilder/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,19 @@
+FROM flynn/cedarish
+
+ADD Gemfile /app/Gemfile
+ADD Gemfile.lock /app/Gemfile.lock
+
+# install Go
+RUN wget https://storage.googleapis.com/golang/go1.4.2.linux-amd64.tar.gz
+RUN tar -C /usr/local -zxvf go1.4.2.linux-amd64.tar.gz
+ENV PATH /usr/local/go/bin:$PATH
+ENV GOPATH /build/gopath
```
Comment (titanous) [inline util/assetbuilder/Dockerfile]: Can we do the build outside of the container via Tup? We already have a Go toolchain available, so this just slows things down.

File: util/assetbuilder/Dockerfile (hunk 2 above)
Comment (jvatic) [inline util/assetbuilder/Dockerfile]: The curl step just adds a repo and still requires running apt-get install.

Comment (titanous) [inline util/assetbuilder/Dockerfile]: Gotcha, that deserves a comment.

**Comment by jvatic** [comment]
@titanous comments addressed

File: dashboard/app/compiler.go [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+func parsePrevManifest() *matrix.Manifest {
+	prevManifest := &matrix.Manifest{
+		Assets: make(map[string]string, 0),
+	}
+	file, err := os.Open("./build/assets/manifest.json")
+	if err != nil {
+		return prevManifest
+	}
+	dec := json.NewDecoder(file)
+	dec.Decode(&prevManifest)
```
Comment (titanous) [inline dashboard/app/compiler.go]: No need to assign `dec`

File: dashboard/app/compiler.go [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+func parsePrevManifest() *matrix.Manifest {
+	prevManifest := &matrix.Manifest{
+		Assets: make(map[string]string, 0),
+	}
+	file, err := os.Open("./build/assets/manifest.json")
+	if err != nil {
+		return prevManifest
+	}
+	dec := json.NewDecoder(file)
```
Comment (titanous) [inline dashboard/app/compiler.go]: `defer file.Close()`

File: dashboard/app/compiler.go [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+	tmpl, err := template.New("dashboard.html").Funcs(template.FuncMap{
+		"assetPath": func(p string) string {
+			return "/assets/" + manifest.Assets[p]
+		},
+	}).Parse(tmplHTML)
+	if err != nil {
+		return err
+	}
+	file, err := os.Create("./build/dashboard.html")
+	defer file.Close()
```
Comment (titanous) [inline dashboard/app/compiler.go]: Always defer close after the error check.

File: dashboard/app/compiler.go [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+	}
+	return tmpl.Execute(file, &TemplateData{
+		Development: os.Getenv("ENVIRONMENT") == "development",
+	})
+}
+
+func readTemplate() (string, error) {
+	var buf bytes.Buffer
+	file, err := os.Open("./lib/dashboard.html.tmpl")
+	defer file.Close()
```
Comment (titanous) [inline dashboard/app/compiler.go]: Ditto.

File: dashboard/app/compiler.go [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+	return buf.String(), nil
+}
+
+func cleanupAssets(prevManifest *matrix.Manifest, manifest *matrix.Manifest) {
+	for logicalPath, path := range prevManifest.Assets {
+		if manifest.Assets[logicalPath] == path {
+			continue
+		}
+		p := filepath.Join("./build/assets", path)
+		if _, err := os.Stat(p); err != nil {
```
Comment (titanous) [inline dashboard/app/compiler.go]: Any reason to do this stat? You don't check the `os.Remove` error.

File: installer/app/compiler.go [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+		if manifest.Assets[logicalPath] == path {
+			continue
+		}
+		p := filepath.Join("./build", path)
+		if _, err := os.Stat(p); err != nil {
+			continue
+		}
+		os.Remove(p)
+	}
+}
```
Comment (titanous) [inline installer/app/compiler.go]: These functions appear duplicated, share from a package?

File: util/assetbuilder/Dockerfile [hunk 10]
```diff
@@ -3,8 +3,11 @@ FROM flynn/cedarish
 ADD Gemfile /app/Gemfile
 ADD Gemfile.lock /app/Gemfile.lock
 
+# Install NodeSource repo
+RUN curl -sL https://deb.nodesource.com/setup | bash -
```
Comment (titanous) [inline util/assetbuilder/Dockerfile]: This run should be merged into the one below.

**Comment by jvatic** [comment]
@titanous comments addressed

**Comment by titanous** [comment]
LGTM

**Pull request merged**